File: JavaScript/src/Internals/Functions/Function_1.js
import { FunctionBase } from './FunctionBase.js';

/**
 * Represents the base class for functions with one parameter
 */
export class Function_1 extends FunctionBase {

    constructor(a) {
        super();
        this.a = a;
    }
    
    /**
     * @param {Array} stringBuilder
     * @param {boolean} addBrackets
     */
    toString2(stringBuilder, addBrackets) {
        this.addFunction(stringBuilder, this.Name);
    }
    
    /**
     * @param {Array} stringBuilder
     * @param {string} functionName
     */
    addFunction(stringBuilder, functionName) {
        stringBuilder.push(functionName);
        stringBuilder.push('(');
        this.a.toString2(stringBuilder, false);
        stringBuilder.push(')');
    }
    
    /**
     * @param {AlgorithmEngine} work
     * @param {Function} tempParameter
     * @returns {Operand}
     */
    getText_1(work, tempParameter) {
        let args1 = this.a.evaluate(work, tempParameter);
        if (args1.IsText) return args1;
        return this.convertToText(args1, 1);
    }
    
    /**
     * @param {AlgorithmEngine} work
     * @param {Function} tempParameter
     * @returns {Operand}
     */
    getNumber_1(work, tempParameter) {
        let args1 = this.a.evaluate(work, tempParameter);
        if (args1.IsNumber) return args1;
        return this.convertToNumber(args1, 1);
    }
    
    /**
     * @param {AlgorithmEngine} work
     * @param {Function} tempParameter
     * @returns {Operand}
     */
    getDate_1(work, tempParameter) {
        let args1 = this.a.evaluate(work, tempParameter);
        if (args1.IsDate) return args1;
        return this.convertToDate(args1, 1);
    }
    
    /**
     * @param {AlgorithmEngine} work
     * @param {Function} tempParameter
     * @returns {Operand}
     */
    getBoolean_1(work, tempParameter) {
        let args1 = this.a.evaluate(work, tempParameter);
        if (args1.IsBoolean) return args1;
        return this.convertToBoolean(args1, 1);
    }
    
    /**
     * @param {AlgorithmEngine} work
     * @param {Function} tempParameter
     * @returns {Operand}
     */
    getArray_1(work, tempParameter) {
        let args1 = this.a.evaluate(work, tempParameter);
        if (args1.IsArray) return args1;
        return this.convertToArray(args1, 1);
    }
    
    evaluate(work, tempParameter = null) {
        throw new Error('FIXME');
    }

}

File: java/src/main/java/toolgood/algorithm/internals/functions/Function_1.java
package toolgood.algorithm.internals.functions;

import java.lang.StringBuilder;
import java.util.List;
import java.util.function.BiFunction;

import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.ParameterType;

public abstract class Function_1 extends FunctionBase {
    protected FunctionBase func1;

    protected Function_1(FunctionBase func1) {
        this.func1 = func1;
    }

    protected Function_1(FunctionBase[] funcs) {
        this.func1 = funcs[0];
    }

    @Override
    public void toString(StringBuilder stringBuilder, boolean addBrackets) {
        AddFunction(stringBuilder, Name());
    }

    protected void AddFunction(StringBuilder stringBuilder, String functionName) {
        stringBuilder.append(functionName);
        stringBuilder.append('(');
        func1.toString(stringBuilder, false);
        stringBuilder.append(')');
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.NONE;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result,
            OperandType operandType, String op, String val) {
    }

    protected Operand GetText_1(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        Operand args1 = func1.Evaluate(engine, tempParameter);
        if (args1.IsText()) return args1;
        return ConvertToText(args1, 1);
    }

    protected Operand GetNumber_1(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        Operand args1 = func1.Evaluate(engine, tempParameter);
        if (args1.IsNumber()) return args1;
        return ConvertToNumber(args1, 1);
    }

    protected Operand GetDate_1(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        Operand args1 = func1.Evaluate(engine, tempParameter);
        if (args1.IsDate()) return args1;
        return ConvertToDate(args1, 1);
    }

    protected Operand GetBoolean_1(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        Operand args1 = func1.Evaluate(engine, tempParameter);
        if (args1.IsBoolean()) return args1;
        return ConvertToBoolean(args1, 1);
    }

    protected Operand GetArray_1(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        Operand args1 = func1.Evaluate(engine, tempParameter);
        if (args1.IsArray()) return args1;
        return ConvertToArray(args1, 1);
    }
}
